swarm_dial.go

```diff
@@ -352,23 +353,66 @@ func (s *Swarm) dial(ctx context.Context, p peer.ID) (*Conn, error) {
 	if len(goodAddrs) == 0 {
 		return nil, &DialError{Peer: p, Cause: ErrNoGoodAddresses}
 	}
-	goodAddrsChan := make(chan ma.Multiaddr, len(goodAddrs))
-	nonBackoff := false
+
+	/////// Check backoff andnRank addresses
+	var nonBackoff bool
 	for _, a := range goodAddrs {
 		// skip addresses in back-off
 		if !s.backf.Backoff(p, a) {
 			nonBackoff = true
-			goodAddrsChan <- a
 		}
 	}
-	close(goodAddrsChan)
 	if !nonBackoff {
 		return nil, ErrDialBackoff
 	}
-	/////////
 
-	// try to get a connection to any addr
-	connC, dialErr := s.dialAddrs(ctx, p, goodAddrsChan)
+	// sorts addresses in descending order of preference for dialing
+	// Private UDP > Public UDP > Private TCP > Public TCP > UDP Relay server > TCP Relay server
+	sortAddrs := func(addrs []ma.Multiaddr) {
+		scoreFnc := func(addr ma.Multiaddr) int {
+			isPrivate := manet.IsPrivateAddr(addr)
+			isFdConsuming := s.IsFdConsumingAddr(addr)
+			_, err := addr.ValueForProtocol(ma.P_CIRCUIT)
+			relay := err == nil
+
+			// UDP
+			if !isFdConsuming {
+				// private UDP
+				if isPrivate {
+					return 1
+				}
+				// public udp
+				if !relay {
+					return 2
+				}
+				// Relay UDP
+				return 5
+			}
+			// It's all TCP now
+			// private TCP
+			if isPrivate {
+				return 3
+			}
+			// public tcp
+			if !relay {
+				return 4
+			}
+			return 6
+		}
```

Comment:
nit: Numeric scores are hard to read and easy to get wrong. Honestly, now that I see this, I'd go back to sorting these addresses into separate slices then concatenating them like you were doing before (sorry for going back and forth on this).

Note: This _is_ fine as-is, just a bit confusing.

Reply:
I agree. I think that way is easier to read. This is done.